pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java

```diff
@@ -874,19 +879,32 @@ protected void doTakeSnapshot() {
         numConsumingSegments++;
         continue;
       }
-      ImmutableSegmentImpl immutableSegment = (ImmutableSegmentImpl) segment;
-      if (!immutableSegment.hasValidDocIdsSnapshotFile()) {
-        segmentsWithoutSnapshot.add(immutableSegment);
+      if (!_updatedSegmentsSinceLastSnapshot.contains(segment)) {
+        // if no updates since last snapshot then skip
         continue;
       }
-      immutableSegment.persistValidDocIdsSnapshot();
-      numImmutableSegments++;
-      numPrimaryKeysInSnapshot += immutableSegment.getValidDocIds().getMutableRoaringBitmap().getCardinality();
+      try {
+        ImmutableSegmentImpl immutableSegment = (ImmutableSegmentImpl) segment;
+        if (!immutableSegment.hasValidDocIdsSnapshotFile()) {
+          segmentsWithoutSnapshot.add(immutableSegment);
+          continue;
+        }
+        immutableSegment.persistValidDocIdsSnapshot();
+        _updatedSegmentsSinceLastSnapshot.remove(segment);
+        numImmutableSegments++;
+        numPrimaryKeysInSnapshot += immutableSegment.getValidDocIds().getMutableRoaringBitmap().getCardinality();
+      } catch (Exception e) {
+        _logger.warn("Caught exception while taking snapshot for segment: {}, skipping", segment.getSegmentName(), e);
```

Comment:
There could be a potential issue to skip failed segments. If we skipped some segments due to failure but have successfully taken snapshot for the latest committed segment, the bitmaps persisted on disk might have valid docs with duplicate PKs. With duplicate PKs, the preloading logic is not ensured to be correct, as it assumes that bitmaps on disk represent a set of valid docs all with unique PKs so that it can simply do Map.put() instead of more costly read-check-write.

If the try-catch here was meant to avoid blocking ingestion from snapshot failure, perhaps better to return early upon failure, instead of skipping. thoughts?

Reply:
Very good point! Ah I see this method: `addSegmentWithoutUpsert` is used in `preloadSegment` flow and uses the `put` method directly. But it seems the ordering is not guaranteed using `ConcurrentHashMap.newKeySet()` ([link](https://javadoc.io/static/net.sf.ehcache/ehcache/2.10.6/net/sf/ehcache/util/concurrent/ConcurrentHashMap.html#newKeySet--:~:text=Because%20the%20elements%20of%20a%20ConcurrentHashMap%20are%20not%20ordered%20in%20any%20particular%20way%2C%20and%20may%20be%20processed%20in%20different%20orders%20in%20different%20parallel%20executions%2C%20the%20correctness%20of%20supplied%20functions%20should%20not%20depend%20on%20any%20ordering)) and we may end up in this situation even in the present scenario or by skipping early. Right?

Reply:
`preloadSegments()` method filter out the segments with bitmap snapshot and then add them with `addSegmentWithoutUpsert()`, so it's important to ensure that those segments' bitmap snapshots as persisted on disk should be disjoint (i.e. not contain duplicate PKs). And with this assumption, the order of loading those segments during this preloading phase doesn't matter. Once preloading finishes, the remaining segments are loaded normally, i.e. doing read-check-write flow.

Reply:
> preloadSegments() method filter out the segments with bitmap snapshot and then add them with addSegmentWithoutUpsert(), so it's important to ensure that those segments' bitmap snapshots as persisted on disk should be disjoint (i.e. not contain duplicate PKs). 

Agreed! So say you had all snapshots taken for all segments previously in one cycle for a table and then in the latest snapshot run, it errored out after processing only few segments. But then even now (not with this change) wouldn't we end up in a similar state?

Reply:
good question, and this method was refactor earlier on to firstly take snapshots for segments already having snapshot on disk, and then take snapshot for those w/o snapshots (i.e. why there are two for-loops in the method), to make sure that even if failure happens in the middle of taking snapshots, the snapshots ever put on disk continue to contain a set of disjoint valid docs.

Reply:
Ah now everything makes sense! Thanks for the clarification! So we follow the principle that whatever is on disk should be disjoint and process in that way itself. Actually that sounds right.
Let me update accordingly!